Product Collection: Add validation for `__experimentalRegisterProductCollection` arguments

imanish003 opened the pull request:
### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Closes #47963

This PR adds validation for arguments of the `registerProductCollection` function. Here are the highlights of this PR:
- It will print a warning in the console for any validation error. It's worth noting that it won't throw an error or bail early.
- It will ignore any extra properties that are not part of the API or aren't allowed. For example, `isDefault`, `isActive` etc. 
  - For this we are [constructing a new object](https://github.com/woocommerce/woocommerce/blob/ae267aa28ec83ab9f33309708d7c3aaf64857065/plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx#L339-L380) with only allowed properties.
- It will validate that all collection names should start with a prefix like [a-zA-Z0-9-]+/product-collection/[a-zA-Z0-9-]+. This will help in avoiding conflicts with other collections from different plugins. But it will just print a warning in the console if the collection name doesn't start with a prefix. It won't throw an error OR bail early.

These warnings will help 3PDs to debug their code and make sure that they are using the API correctly.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

> [!IMPORTANT]
> To be tested by developers only.

1. Download following plugin, install and activate it: [register-product-collection-tester.zip](https://github.com/user-attachments/files/15633348/register-product-collection-tester.zip)
2. You will see that there are three collections registered in `collections.js` file using `__experimentalRegisterProductCollection` function:
   - My Custom Collection
   - My Custom Collection with Preview
   - My Custom Collection with Advanced Preview
3. Change various arguments passed to `__experimentalRegisterProductCollection` function in `collections.js` file and see if you are getting warning in console for validation errors, when you refresh the Editor page.
    - You can see all the validations in [this file](https://github.com/woocommerce/woocommerce/blob/ae267aa28ec83ab9f33309708d7c3aaf64857065/plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx). This will help you understand which arguments are validated & message you will see in console for each validation error.

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box and supplying data. -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [x] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [x] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->
Add validation for `__experimentalRegisterProductCollection` arguments

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

imanish003 committed:
Add validation for `__experimentalRegisterProductCollection` arguments

Introduced comprehensive validation for the `ProductCollectionConfig` object in `__experimentalRegisterProductCollection` to ensure correct data types and values, enhancing error handling and robustness.

- Added a new function `isValidProductCollectionConfig` to perform various checks on the `ProductCollectionConfig` object.
- Validates properties such as `name`, `title`, `description`, `category`, `keywords`, `icon`, `isDefault`, `innerBlocks`, `example`, `scope`, `isActive`, `attributes`, and `preview`.
- Ensures correct data types and provides detailed console error messages for invalid configurations.
- Updated `__experimentalRegisterProductCollection` to use the validation function before proceeding with the registration process.

**Impact**
- Improves stability and prevents invalid configurations from causing runtime errors.
- Provides clearer error messages for developers, aiding in quicker debugging and development.

imanish003 commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
```diff
@@ -26,17 +27,364 @@ export interface ProductCollectionConfig extends BlockVariation {
+		return false;
+	}
+
+	// BlockVariation properties
+	if ( typeof config.name !== 'string' || config.name.length === 0 ) {
+		console.error( 'Invalid name: name must be a non-empty string.' );
+		return false;
+	}
+	if (
+		config.title !== undefined &&
+		( typeof config.title !== 'string' || config.title.length === 0 )
+	) {
+		console.error( 'Invalid title: title must be a non-empty string.' );
+		return false;
+	}
```
**Question**: Should we add validation for core arguments of `registerBlockVariation` function from `@wordpress/blocks` package? Maybe we should add validation for all arguments we are passing to `registerBlockVariation` function, because Gutenberg does not have enough validation 🤔 

**More info:**

Here is how we import `registerBlockVariation` function:
```
import { registerBlockVariation } from '@wordpress/blocks';
```

It does not seem to have any validation for arguments, as you can see [here](https://github.com/WordPress/gutenberg/blob/13717c6d83b3548fc967f1006cffb09093e1faec/packages/blocks/src/api/registration.js#L725-L731):
```tsx
export const registerBlockVariation = ( blockName, variation ) => {
	if ( typeof variation.name !== 'string' ) {
		console.warn( 'Variation names must be unique strings.' );
	}

	dispatch( blocksStore ).addBlockVariations( blockName, variation );
};
```

Which calls [following action](https://github.com/WordPress/gutenberg/blob/13717c6d83b3548fc967f1006cffb09093e1faec/packages/blocks/src/store/actions.js#L146-L152) to add the variation:
```tsx
export function addBlockVariations( blockName, variations ) {
	return {
		type: 'ADD_BLOCK_VARIATIONS',
		variations: Array.isArray( variations ) ? variations : [ variations ],
		blockName,
	};
}
```

Which then has following reducer to add the variation:
```tsx
case 'ADD_BLOCK_VARIATIONS':
	return {
		...state,
		[ action.blockName ]: getUniqueItemsByName( [
			...( state[ action.blockName ] ?? [] ),
			...action.variations,
		] ),
	};
```

It doesn't have any validation for arguments. It only checks if `variation.name` is a string or not.
- It adds the object key as blockName and value as an array of variations.

One thing to note is that it uses `getUniqueItemsByName` function, which looks like [this](https://github.com/WordPress/gutenberg/blob/13717c6d83b3548fc967f1006cffb09093e1faec/packages/blocks/src/store/reducer.js#L51-L58):
```tsx
function getUniqueItemsByName( items ) {
	return items.reduce( ( acc, currentItem ) => {
		if ( ! acc.some( ( item ) => item.name === currentItem.name ) ) {
			acc.push( currentItem );
		}
		return acc;
	}, [] );
}
```

So basically, if there are multiple variations with the same name, it will only keep one of them. CC: @xristos3490 I think we talked about this in one of our chats.

**Another Question:** I am not 100% sure why arguments are not validated in Gutenberg. Could it be for performance reasons?

kmanijak replied:
> Should we add validation for core arguments of registerBlockVariation function from @wordpress/blocks package?

Hmm.. Considering the API is in use long time (widely and successfully) I don't think we have to be more strict than Gutenberg but obviously is nice-to-have. I wouldn't remove it if we have it already!

imanish003 replied:
Sounds good! As I have already added validation for core arguments of registerBlockVariation function, I will keep it as it is :handshake:

imanish003 commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
**Question:** Should we consider using a library for validation? I can see that we are using `zod` [here](https://github.com/woocommerce/woocommerce/blob/a4544afa363a37dd6d9ca572f275b57031c21215/plugins/woocommerce-admin/client/core-profiler/pages/BusinessInfo.tsx#L22) in `woocommerce-admin`. Can't we use it for validation in WooCommerce Blocks? Although, I can also use JavaScript to validate the arguments but I think using a library will be better.

I just want to know if there is a good reason to not use a library for validation in WooCommerce Blocks.

imanish003 commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
```diff
@@ -26,17 +27,364 @@ export interface ProductCollectionConfig extends BlockVariation {
+		console.error( 'Invalid featured: featured must be a boolean.' );
+		return false;
+	}
+	// attributes.query.timeFrame
+	if (
+		config.attributes?.query?.timeFrame !== undefined &&
+		typeof config.attributes.query.timeFrame !== 'object'
+	) {
+		console.error( 'Invalid timeFrame: timeFrame must be an object.' );
+		return false;
+	}
```
**Question:** How deep should we go for validation? For example, when I am validating `attributes.query.timeFrame`. Should I validate `operator` and `value` properties of `TimeFrame` object?

Here are some more such examples:
- woocommerceStockStatus: Should I validate if each array item is a valid value?
- query.priceRange: Should I validate if `min` and `max` are numbers?
- displayLayout: Should I validate if `type` is a valid value from `LayoutOptions` enum? Should I validate if `columns` is a number and `shrinkColumns` is a boolean?
- query.order: Should I validate if it's a valid value i.e. 'asc' or 'desc'?
- query.orderBy: Should I validate if it's a valid value i.e. 'date', 'title', 'popularity', 'rating'?

There could be other such properties too. I just wanna understand sentiment on how deep should we go for validation. Based on that I can update the PR. Looking forward to your thoughts.

kmanijak replied:
I don't think that's necessary.

kmanijak replied:
I don't think that's necessary.

imanish003 replied:
Sounds good! Let's keep validation top-level only. If needed, we will make it more strict in future.  CC: @xristos3490

imanish003 commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
**Question**: What's purpose of `queryContext` in attributes? Looks like it's being used to set only page number as can be seen [here](https://github.com/woocommerce/woocommerce/blob/a4544afa363a37dd6d9ca572f275b57031c21215/plugins/woocommerce-blocks/assets/js/blocks/product-template/edit.tsx#L152). Can it be deprecated? Is there a good reason to keep it?

imanish003 commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
**Question**: Should we allow setting `forcePageReload` in `attributes`? IMO We should keep it for internal use only for now. Later we may reconsider it. What do you think?

kmanijak replied:
I agree, it should be kept internal.

imanish003 commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
```diff
@@ -26,17 +27,364 @@ export interface ProductCollectionConfig extends BlockVariation {
+	}
+
+	// BlockVariation properties
+	if ( typeof config.name !== 'string' || config.name.length === 0 ) {
+		console.error( 'Invalid name: name must be a non-empty string.' );
+		return false;
+	}
```
**Question**: Should we add validation such that all collection names should start with a prefix? For example, `[my-plugin-name]/product-collection/[my-collection-name]`. We can have a regex to check if the collection name starts with a prefix like `[a-zA-Z0-9-]+/product-collection/[a-zA-Z0-9-]+`. What do you think?
- This will help in avoiding conflicts with other collections from different plugins.

I had chat with @xristos3490 about this. Just wanted to raise this question here for more input & discussion.

kmanijak replied:
I'm not sure if such strict validation is required... But I'd at least restrict starting with `woocommerce/` - IMO that should be forbidden.

imanish003 replied:
I don't have strong opinion on this. My rational behind using plugin name prefix is to avoid conflicts with other collections from different plugins. But If you think it's not required then I am happy to go with that. I will wait for @xristos3490 input on this before making any decision.

imanish003 reviewed (commented):
Hi @kmanijak @xristos3490,

I have some questions regarding the validation of arguments. I have added those questions as comments on the PR. I would appreciate your thoughts on those questions. Looking forward to your feedback. Thank you!

imanish003 committed:
Fix typo

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
```diff
@@ -26,17 +27,364 @@ export interface ProductCollectionConfig extends BlockVariation {
+	}
+	if (
+		config.isDefault !== undefined &&
+		typeof config.isDefault !== 'boolean'
```
We don't allow providing `isDefault`, so we should rather completely ignore the property.

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
```diff
@@ -26,17 +27,364 @@ export interface ProductCollectionConfig extends BlockVariation {
+		config.isActive !== undefined &&
+		typeof config.isActive !== 'function'
+	) {
+		console.error( 'Invalid isActive: isActive must be a function.' );
```
We agreed `isActive` is not a subject to change:

> You can see all other supported arguments in the [Block Variation](https://developer.wordpress.org/block-editor/reference-guides/block-api/block-variations/#defining-a-block-variation) documentation. You won't be able to use isDefault attribute as we are setting it to false for all 3PDs collections. Also, isActive can't be used as it will be managed by us.

from https://github.com/woocommerce/woocommerce/pull/48141 description, hence property should be ignored completely.

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
```diff
@@ -26,17 +27,364 @@ export interface ProductCollectionConfig extends BlockVariation {
+	// attributes.queryId - queryId is generated automatically
+	if ( config.attributes?.queryId !== undefined ) {
+		console.error(
+			'Invalid queryId: queryId must not be set as it will be generated automatically.'
```
Nice one! 👏

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
```diff
@@ -26,17 +27,364 @@ export interface ProductCollectionConfig extends BlockVariation {
+		);
+		return false;
+	}
+	// attributes.collection - It will be set to args.name
```
I'd suggest:

```suggestion
	// attributes.collection - It will be derived from args.name
```

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
```diff
@@ -26,17 +27,364 @@ export interface ProductCollectionConfig extends BlockVariation {
+	// attributes.collection - It will be set to args.name
+	if ( config.attributes?.collection !== undefined ) {
+		console.error(
+			'Invalid collection: collection must not be set as it will be set automatically to config.name.'
```
```suggestion
			'Invalid collection: collection must not be set as it will be automatically derived from config.name.'
```

kmanijak reviewed (changes requested):
I left some comments but also have a general suggestion that we should not throw an error when properties that we ignore are not provided.

What I mean is for example `forcePageReload`.

I suggest something like:
```
	// attributes.forcePageReload - For internal use only
	if ( config.attributes?.forcePageReload !== undefined ) {
		console.warn(
			'forcePageReload property will be ignored.'
		);
	}
```

- warning instead of error - collection can still be registered, it's just property will be ignored
- don't bail early (no `return false`) - collection will still work correctly.

Alternatively, we can also ignore these properties in the same way as if a user provided some additional properties, they will be simply ignored. Imagine this:

```
wc.wcBlocksRegistry.__experimentalRegisterProductCollection({
  name: 'My custom Collection',
  ...
  randomKey: 'invalid but will be ignored',
  // in the same way what is a part of Variation API IS NOT a part of the API,
  // so I don't think we even have to notify 3PDs that they provide extra stuff.
  // We should just ignore `isDefault` and other properties.
  isDefault: true 
})
```

So the above applies to:
- `isDefault`
- `isActive`
- `attributes.query.exclude` - I don't think we use that property atm 🤔 
- `attributes.query.queryId`
- `attributes.convertedFromProducts`
- `attributes.collection`
- `attributes.forcePageReload`
- `attributes.__privatePreviewState`

And if we apply the latter solution (which I'm in favor for!) we should simply pick specific properties from a config at `__experimentalRegisterProductCollection` level and omit all the extra properties that are not part of the API.

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce into add/47963-product-collection-add-validation-for-registerproductcollection-arguments

imanish003 committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce into add/47963-product-collection-add-validation-for-registerproductcollection-arguments

imanish003 committed:
Merge branch 'add/47962-product-collection-expose-experimental-registerproductcollection-for-3pds' of https://github.com/woocommerce/woocommerce into add/47963-product-collection-add-validation-for-registerproductcollection-arguments

imanish003 commented:
> And if we apply the latter solution (which I'm in favor for!) we should simply pick specific properties from a config at __experimentalRegisterProductCollection level and omit all the extra properties that are not part of the API.

@kmanijak  I like this idea. I will make the changes accordingly. Thanks for the suggestion! :bow:

imanish003 committed:
Refactor: Replace console.error with console.warn

Updated the error logging in the isValidProductCollectionConfig function to use console.warn instead of console.error for invalid configuration properties. This address the feedback from the PR review.

- Replaced console.error with console.warn for various validation checks in isValidProductCollectionConfig.
- Removed redundant return statements after console.warn calls.
- Improved logging messages to better inform about invalid configuration properties without treating them as critical errors.
- Simplified the logic in __experimentalRegisterProductCollection by combining query and attribute properties and ensuring defaults are set properly.

imanish003 committed:
Refactor: Rename isValidProductCollectionConfig to isValidCollectionConfig

Updated the function name from isValidProductCollectionConfig to isValidCollectionConfig for better clarity and consistency. Also, renamed related variables for improved readability.

imanish003 committed:
Add validation for name property

Commit:
Add changefile(s) from automation for the following project(s): woocommerce-blocks

imanish003 commented:
Hey @kmanijak! I have made the changes and PR is ready for review. I have updated the PR description with the changes I have made. I think I have addressed the comments you left on the PR. Please let me know in case I missed anything. I am looking forward to hearing your thoughts on this 🙇 

As I communicated in Slack(p1719292878030069/1718871438.463229-slack-C06N9CP18BU), I planned to use AJV for validation, but there is one limitation. The limitation is that it doesn't allow the "Custom Error Messages" setting for each validation rule. We need to use another library called [ajv-errors](https://ajv.js.org/packages/ajv-errors.html) in order to set custom error messages. I am not sure if we want to use another library for this. Therefore:
- I will use JavaScript for validation for now.
- I will share my findings on [Github Discussion](https://github.com/woocommerce/woocommerce/discussions/47277). Once we have made a decision about which library to use, we can switch to that library.

imanish003 committed:
Title is required for new collection

imanish003 committed:
Merge branch 'add/47963-product-collection-add-validation-for-registerproductcollection-arguments' of https://github.com/woocommerce/woocommerce into add/47963-product-collection-add-validation-for-registerproductcollection-arguments

imanish003 committed:
Update comments

imanish003 committed:
Merge branch 'add/47962-product-collection-expose-experimental-registerproductcollection-for-3pds' of https://github.com/woocommerce/woocommerce into add/47963-product-collection-add-validation-for-registerproductcollection-arguments

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
```diff
@@ -26,17 +27,359 @@ export interface ProductCollectionConfig extends BlockVariation {
+	if ( typeof config !== 'object' || config === null ) {
+		console.warn(
+			'Invalid arguments: You must pass an object to __experimentalRegisterProductCollection.'
+		);
```
I see you changed all of the errors to warns. If that was based on my comment:

>…that we should not throw an error when properties that we ignore are not provided.
> - warning instead of error - collection can still be registered, it’s just property will be ignored
> - don’t bail early (no return false) - collection will still work correctly.

then we should get back to `errors` in some cases like this one. I only meant the specific case when someone provides the property that we ignore (because in those cases registering would still happen). In some cases, like this one, we should throw an error and bail early, because we cannot successfully register collection without config or if config is not an object.

Sorry for not explaining this better or highlighting it stronger.

```suggestion
		console.error(
			'Invalid arguments: You must pass an object to __experimentalRegisterProductCollection.'
		);
		return false;
```

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
```diff
@@ -26,17 +27,359 @@ export interface ProductCollectionConfig extends BlockVariation {
+	 */
+	// name
+	if ( typeof config.name !== 'string' || config.name.length === 0 ) {
+		console.warn( 'Invalid name: name must be a non-empty string.' );
```
As above, I think we should use error and bail early when the property is required

```suggestion
		console.error( 'Invalid name: name must be a non-empty string.' );
		return false;
```

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
```diff
@@ -26,17 +27,359 @@ export interface ProductCollectionConfig extends BlockVariation {
+			/^[a-zA-Z0-9-]+\/product-collection\/[a-zA-Z0-9-]+$/
+		)
+	) {
+		console.warn(
```
For example in this case I'd keep warn. I mean, collection doesn't follow the exact convention and it may be overridden by other collection but in general it doesn't prevent successful registration. Let me know if you agree!

imanish003 replied:
Agreed! We don't need to bail early as this is a recommendation to avoid conflicts with other collections from different plugins. But it won't prevent successful registration.

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
```diff
@@ -26,17 +27,359 @@ export interface ProductCollectionConfig extends BlockVariation {
+	) {
+		console.warn( 'Invalid priceRange: priceRange must be an object.' );
+	}
+	// attributes.queryId - queryId is generated automatically
+	if ( config.attributes?.queryId !== undefined ) {
+		console.warn(
+			'Invalid queryId: queryId must not be set as it will be generated automatically.'
+		);
+	}
```
It can be removed as property is ignored anyway
```suggestion
```

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
```diff
@@ -26,17 +27,359 @@ export interface ProductCollectionConfig extends BlockVariation {
+			'Invalid displayLayout: displayLayout must be an object.'
+		);
+	}
+	// attributes.collection - It will be derived from args.name
+	if ( config.attributes?.collection !== undefined ) {
+		console.error(
+			'Invalid collection: collection must not be set as it will be automatically derived from config.name.'
+		);
+		return false;
+	}
```
It can be removed as property is ignored anyway

```suggestion
```

kmanijak commented on plugins/woocommerce-blocks/assets/js/blocks-registry/product-collection/register-product-collection.tsx:
```diff
@@ -26,17 +27,359 @@ export interface ProductCollectionConfig extends BlockVariation {
+	config: ProductCollectionConfig
+) => {
+	// Check if the config is valid. It will log warnings in the console if the config is invalid.
+	isValidCollectionConfig( config );
```
I don't remember if that was the case but if you agree with the bailing early strategy, I think we can also prevent registration if the validation doesn't pass. Something among those lines.

```suggestion
	const isConfigValid = isValidCollectionConfig( config );
	
	if (! isConfigValid) {
	    console.error( `Collection ${config.name} could not be registered due to validation error.` )
	    return false;
	}
```

If you decide to follow that, `isValidCollectionConfig` should return `true` at the end 🙌

imanish003 replied:
Makes sense! I will make these changes.

kmanijak reviewed (changes requested):
I left some comments mainly regarding switching `errors` to `warns` and bailing early. It was miscommunication, apologies that my comment was not clear enough 🙌

imanish003 committed:
Fix E2E tests

imanish003 committed:
Address PR feedback

imanish003 commented:
Hey @kmanijak! I have addressed the feedback you have left on the PR. I have made the changes accordingly. Can you please review PR again when you have some time?

kmanijak reviewed (approved):
Thanks for patience and addressing all of the comments and questions! Looks good! 🚀
